straxen/plugins/s2cnn_processing.py

```diff
@@ -0,0 +1,154 @@
+                               )[0][0]
+            x_offset = ( 2*int(np.ceil( 
+                           0.5* (row_nPMT.max()  - row_nPMT[n_row+1]))) 
+                         - (row_nPMT[n_row+1]+1)%2)
+            i_PMT = i - row_nPMT_cumulative[n_row]
+            x_coord = x_offset + 2*i_PMT 
+            y_coord = tot_rows - n_row
+            coords[i] = [x_coord, y_coord]
+            pairs.append( (x_coord, y_coord))
+        return(coords,np.array(pairs) )
+    
+    def __init__(self):
+        self.coords,self.pairs = self.PMTnumuber_doublewidth()
+        self.size = (33,19)
+        
+    def get_coordinates(self):
+        """
+        this function returns dictionary of cooridnates
+        if form dict(ch: [x,y]) 
+        """
+        return(deepcopy(self.coords))
+    
+    def get_size(self):
+        """ 
+        return size of the 2D image 
+        """
+        return(deepcopy(self.size))
+    
+    def convert_pattern(self, inarr):
+        """
+        This function converts an array of PMT areas to 2D pattern
+        """
+        pattern = np.zeros(self.size)
+        pattern[self.pairs[:,0],self.pairs[:,1] ] = np.array(inarr)
+        return(pattern)
+    
+    def convert_multiple_patterns(self, inarr):
+        """
+        This function converts a 2D array of PMT arreas (for multiple events)
+        """
+        pattern = np.zeros((inarr.shape[0],self.size[0],self.size[1]))
+        pattern[:,self.pairs[:,0],self.pairs[:,1] ] = np.array(inarr)
+        return(pattern)
+
+if straxfound:
+    export, __all__ = strax.exporter()
+    @export
+    @strax.takes_config(
+        strax.Option('min_reconstruction_area',
+                     help='Skip reconstruction if area (PE) is less than this',
+                     default=10),
+        strax.Option('n_top_pmts', default=straxen.n_top_pmts,
+                     help="Number of top PMTs"),
+        strax.Option("s2_cnn_model_path", 
+                     help="Path to the CNN model in hdf5 format. WARING, this should include the whole model file and not the weights file", 
+                     default=("/project2/lgrandi/terliuk/CNNmodels/CNN_defaults/disk_const_PE/CNN_dw_3L_cm_maxnorm_A_const_100__Z_const_-1.hdf5")
+                    )
+    )
+    class CNNS2PosRec(strax.Plugin):
+        """
+        This pluging provides S2 position reconstruction for top array
+        
+        returns variables : 
+            - x_TFS2CNN - reconstructed X position in [ cm ]
+            - y_TFS2CNN - reconstructed Y position in [ cm ]
+            - patterns - 2D array of normalized areas as used for CNN
+        """
+        dtype = [('x_TFS2CNN', np.float32,
+                  'Reconstructed CNN S2 X position [ cm ] '),
+                 ('y_TFS2CNN', np.float32,
+                  'Reconstructed CNN S2 Y position [ cm ] '), 
+                 (("Patterns after DW transformation normalized to max PMT area",
+                     "patterns"), np.float, (33, 19,)), 
```

Comment:
Better make this an option?

Reply:
Do you mean storage of patterns?

Reply:
I added a switch that defines whether patterns are stored or not in https://github.com/XENONnT/straxen/pull/320/commits/115a4acbeb3d30532d76523e290b8a5bbe056732